src/couch_srt/src/couch_srt_query.erl

```diff
@@ -0,0 +1,985 @@
+    group_by(Matcher, KeyFun, ValFun, AggFun, query_cardinality_limit()).
+
+-spec all() ->
+    matcher().
+
+all() ->
+    Spec = ets:fun2ms(fun(#rctx{} = R) -> R end),
+    {Spec, ets:match_spec_compile(Spec)}.
+
+%% eg: group_by(all(), username, docs_read).
+%% eg: ^^ or: group_by(all(), [username, docs_read], ioq_calls).
+%% eg: group_by(all(), [username, dbname, js_filter], docs_read).
+%% eg: group_by(all(), [username, dbname, js_filter], ioq_calls).
+%% eg: group_by(all(), [username, dbname, js_filter], get_kv_node).
+-spec group_by(Matcher, KeyFun, ValFun, AggFun, Limit) ->
+    {ok, aggregation_result()} | {limit, aggregation_result()}
+when
+    Matcher :: matcher(),
+    KeyFun ::
+        aggregation_keys_fun()
+        | rctx_field()
+        | [rctx_field()],
+    ValFun ::
+        value_key_fun()
+        | rctx_field(),
+    AggFun ::
+        count_key_fun(),
+    Limit :: pos_integer().
+
+group_by(Matcher, KeyL, ValFun, AggFun, Limit) when is_list(KeyL) ->
+    KeyFun = fun(Ele) -> list_to_tuple([couch_srt_entry:value(Key, Ele) || Key <- KeyL]) end,
+    group_by(Matcher, KeyFun, ValFun, AggFun, Limit);
+group_by(Matcher, Key, ValFun, AggFun, Limit) when is_atom(Key) ->
+    group_by(Matcher, curry_field(Key), ValFun, AggFun, Limit);
+group_by(Matcher, KeyFun, Val, AggFun, Limit) when is_atom(Val) ->
+    group_by(Matcher, KeyFun, curry_field(Val), AggFun, Limit);
+group_by(Matcher, KeyFun, ValFun, AggFun, Limit) ->
+    %% This is a space versus speed tradeoff. Both query modes only filter
+    %% through the table until `Limit` rows have been returned and both will
+    %% utilize the compiled match_specs to do the testing, but
+    %% `group_by_fold/5` will sequentially copy in every row and test it
+    %% locally against the compiled match_spec using `ets:foldl/3`, whereas
+    %% `group_by_select/5` does the filtering internally in the ETS NIF, by way
+    %% of passing the uncompiled match_spec to `ets:select/3` to
+    %% `ets:select/3`. The tradeoff here is that `ets:select` will copy `Limit`
+    %% full `#rctx{}` records into this caller process, which we then aggregate
+    %% over, as opposed to `ets:foldl` only sequentially loading a singular
+    %% `#rctx{}` and extracting the relevant field value to aggregate on.
+    %%
+    %% The use of `query_by_fold` should only be needed if `Limit` is
+    %% drastically increased, and even then, the efficiencies gained here with
+    %% `query_by_fold` are shortlived until we can encode the values needed by
+    %% `ValFun` into the match_spec return fields, at which point it becomes
+    %% strictly worse.
+    %%
+    %% NOTE: This discrepancy of `ets:match_spec_run` taking a `match_spec()`
+    %% vs `ets:select` taking a `comp_match_spec()` is why our CSRT `matcher()`
+    %% type_spec funnels around both versions instead of just reference to the
+    %% compiled spec stored by ETS internally.
+    case config:get_boolean(?CSRT, "use_query_fold", false) of
```

Comment:
How large of difference does each method have compared to another? Instead of keeping both around wonder if there is a middle-ground of querying in smaller batches of limit = 500 and doing a bunch of them in a row until we reach the users's `limit`. It just seems like a lot of complexity added for a method that might be called once in a while by a single operator debugging or reporting a large io usage or investigating some performance issues?